test/e2e/framework/util.go

```diff
@@ -1284,3 +1285,110 @@ func taintExists(taints []v1.Taint, taintToFind *v1.Taint) bool {
 	}
 	return false
 }
+
+// WatchEventSequenceVerifier ...
+// manages a watch for a given resource, ensures that events take place in a given order, retries the test on failure
+//   testContext         cancelation signal across API boundries, e.g: context.TODO()
+//   dc                  sets up a client to the API
+//   resourceType        specify the type of resource
+//   namespace           select a namespace
+//   resourceName        the name of the given resource
+//   listOptions         options used to find the resource, recommended to use listOptions.labelSelector
+//   expectedWatchEvents array of events which are expected to occur
+//   scenario            the function to run
+func WatchEventSequenceVerifier(testContext context.Context, dc dynamic.Interface, resourceType schema.GroupVersionResource, namespace string, resourceName string, listOptions metav1.ListOptions, expectedWatchEvents []watch.Event, scenario func(*watchtools.RetryWatcher) []watch.Event) {
+	initResource, err := dc.Resource(resourceType).Namespace(namespace).List(testContext, listOptions)
+	ExpectNoError(err, "Failed to fetch initial resource")
+	listWatcher := &cache.ListWatch{
+		WatchFunc: func(listOptions metav1.ListOptions) (watch.Interface, error) {
+			return dc.Resource(resourceType).Namespace(namespace).Watch(testContext, listOptions)
+		},
+	}
+	resourceWatch, err := watchtools.NewRetryWatcher(initResource.GetResourceVersion(), listWatcher)
+	ExpectNoError(err, "Failed to create a resource watch of %v in namespace %v", resourceType.Resource, namespace)
+
+	// NOTE value of 3 retries seems to make sense
+	retries := 3
+retriesLoop:
+	for try := 1; try <= retries; try++ {
+		// NOTE the test may need access to the events to see what's going on, such as a change in status
+		actualWatchEvents := scenario(resourceWatch)
+		errs := sets.NewString()
+		// watchEventsLoop:
+		totalValidWatchEvents := 0
+		actualWatchEventsHasDelete := false
+		for watchEventIndex := range expectedWatchEvents {
+			foundExpectedWatchEvent := false
+			ExpectEqual(len(expectedWatchEvents) <= len(actualWatchEvents), true, "Error: actual watch events amount must be greater than or equal to expected watch events amount")
+			for actualWatchEventIndex := range actualWatchEvents {
+				if actualWatchEvents[watchEventIndex].Type == expectedWatchEvents[actualWatchEventIndex].Type {
+					foundExpectedWatchEvent = true
+				}
+				if actualWatchEvents[actualWatchEventIndex].Type == watch.Deleted && actualWatchEventsHasDelete != true {
+					actualWatchEventsHasDelete = true
+				}
+			}
+			if foundExpectedWatchEvent == false {
+				errs.Insert(fmt.Sprintf("Watch event %v not found", expectedWatchEvents[watchEventIndex].Type))
+			}
+			totalValidWatchEvents++
+		}
+		if actualWatchEventsHasDelete == false {
+			_ = dc.Resource(resourceType).Namespace(namespace).DeleteCollection(testContext, metav1.DeleteOptions{}, listOptions)
+		}
```

Comment:
??? why are we deleting things here? this function should just be checking event order invariants, right?

Reply:
If retries without deleting things, then on the next retry it will likely fail (e.g: create resource)

Reply:
The name though makes you think it is non-destructive. This is assuming a lot. Perhaps a `cleanup` or `retryCleanup` callback method, which the caller can specify?

Reply:
That's a better idea than what I put it - seems rather flexible.